dolosse/run_control/cli/kafka_pyinq.py

```diff
@@ -0,0 +1,263 @@
+        msg = consumer.poll(timeout=1.0)
+        if msg is None:
+            continue
+        if msg.error():
+            raise KafkaException(msg.error())
+        else:
+            # Proper message
+            sys.stderr.write('%% %s [%d] at offset %d with key %s:\n' %
+                             (msg.topic(), msg.partition(), msg.offset(), str(msg.key())))
+            json_data = loads(msg.value())
+            if json_data['category'] == 'feedback':
+                status['daq_state'] = json_data['daq']['run_state']
+                status['daq_feedback'] = json_data['os']
+                status['daq_evt_rate'] = str(json_data['daq']['evt_rate'])
+                status['daq_kB_rate'] = str(json_data['daq']['kB_rate'])
+                status['daq_events'] = str(json_data['daq']['events'])
+            elif json_data['category'] == 'errors':
+                status['daq_error'] = json_data['msg']
+
+    consumer.close()
+
+
+def read_yaml_config(conf_dict):
+    # read yaml file
+    with open('kafka_interface.yaml') as yf:
+        config = safe_load(yf)
+    conf_dict['bootstrap_servers'] = config['kafka']['bootstrap_servers']
+    conf_dict['topic_control'] = config['kafka']['topics']['control']
+    conf_dict['topics'] = [config['kafka']['topics']['feedback'], config['kafka']['topics']['errors']]
+    conf_dict['group_id'] = config['kafka']['group_id_interface']
+    conf_dict['topic_manage'] = config['kafka']['topics']['manage']
+
+
+if __name__ == '__main__':
+    # start in a state to begin execution, and set DAQ state to "normal"
+    run = True
+
+    # status parameters
+    daq_status = {
+        'daq_state': 'Stopped',
+        'daq_error': ' ',
+        'daq_feedback': ' ',
+        'daq_evt_rate': '0',
+        'daq_kB_rate': '0',
+        'daq_events': '0'
+    }
+
+    # yaml parameters
+    yaml_conf = {
+        'bootstrap_servers': ' ',
+        'topic_control': ' ',
+        'topics': [],
+        'group_id': ' ',
+        'topic_manage': ' '
+    }
+
+    read_yaml_config(yaml_conf)
+
+    # configure producer
+    kafka_conf = {'bootstrap.servers': yaml_conf['bootstrap_servers']}
+
+    # create producer
+    producer = Producer(**kafka_conf)
+
+    # consumer configuration
+    kafka_cons_conf = {'bootstrap.servers': yaml_conf['bootstrap_servers'], 'group.id': yaml_conf['group_id'],
+                       'session.timeout.ms': 6000,
+                       'auto.offset.reset': 'latest'}
+
+    # create consumer
+    consumer = Consumer(kafka_cons_conf)
+    try:
+        consumer.subscribe(yaml_conf['topics'])
+    except KafkaException:
+        print('Kafka Error in subscribing to consumer topics')
+        run = False
+    except RuntimeError:
+        print('Could not subscribe to consumer topics - Consumer closed')
+        run = False
+
+    t2 = Thread(target=command_interface, args=(daq_status, ))
+    t1 = Thread(target=status_readout, args=(daq_status, ))
+
+    t1.start()
+    t2.start()

```

Comment:
Because these are using threading, they're still actually running sequentially (i.e. NOT in parallel). Only one of the threads run at a time because they share the same [GIL](https://wiki.python.org/moin/GlobalInterpreterLock). To make these truly parallel you should use `multiprocessor` instead of `threading`. Here's a [nice high level article](https://medium.com/contentsquare-engineering-blog/multithreading-vs-multiprocessing-in-python-ece023ad55a) about it. 

You're also missing a join here for clean exit from the program (includes a final logging statement)

Reply:
Added join. The threads just need to be concurrent, so the one processor is fine. At some point I did try multiprocessing and it would not work with consumer/producer code. Have you got this working?

Reply:
I have, but would have to hunt up the code on my work computer. If you've got it working roll with it.